### PR TITLE
dasLLAMA: EPYC 9654 campaign — team dispatch, fused decode chains, hybrid backend, dispatch shaping

### PR DESCRIPTION
**Pull request by borisbat**
The EPYC 9654 (96C/192T Genoa) rental campaign: five measurement sessions chasing Llama-1B decode from 7.6 t/s at defaults to a 124 t/s peak — 89% of llama.cpp at the T=24 bandwidth knee — plus a fleet-wide +20% prefill win from the batch-backend pin. Full methodology, per-session tables, and negative results in `modules/dasLLAMA/epyc9654_measurements.md`.

## jobque core

- **Multi-stage team dispatch** (`teamParallelForStages`): one team rendezvous runs N dependent stages with worker-side barriers between them, so a fused norm→matmul→matmul chain pays the dispatch tail once instead of per op. Seqlock publish (odd seq) + an in-flight entrant gate make the inter-chain stage-order invariant hold by construction; per-stage claim words keep the packed {K|ctr} epoch-race fix. Single-op `teamParallelFor` is now `stages(1)` — one protocol. das binding `team_parallel_stages` (jobque_boost) clones the worker lambda once per chain; stage-order storm test in `tests/jobque` (TSAN 25× clean).
- **Team dispatch is the runner default** — fifo fork/join costs ~10µs per job (~1ms per 96-lane dispatch, ~160 dispatches per decoded token = the entire decode budget); the team path measures orders cheaper (decode 5.7×).
- **`get_total_hw_cores()` builtin + `JobQue::set_default_threads_cap()`** — dasLLAMA `[init]` caps its jobque at physical cores − 1 (SMT siblings share FMA/load ports); `min()` with the stock rule preserves the Apple P-core default; `DAS_JOBQUE_THREADS` wins.
- **Dispatch anatomy profiler** (`DAS_TEAM_PROF=1`): per-op publish / caller-serve / join-tail / worker-claim timestamps dumped at queue destruction. It settled the decode dispatch question — the team rendezvous is already barrier-cheap (sub-µs publish, ~200ns worker reaction), closing protocol surgery as a dead end.

## dasLLAMA dispatch shaping

- **ggml-parity min-grain chunking**: 16-row min grain at all matmul dispatch sites (a fixed `lanes*oversplit` chunk count shattered small matmuls at 192 lanes), 64-row grain for GEMV-shaped sites, equal-split cap at one chunk per lane (ggml's decode fallback shape). Both grain knobs ride the box_profile runtime rail.
- **GEMV wave fill** (Boris's rule): a grain-derived chunk count in [lanes/2, lanes) rounds UP to exactly one full wave — chunks shrink but stay ≥ grain/2, no lane idles. ABBA @24 lanes: ~+14% decode. Knob `set_gemv_wave_fill` (default ON).
- **Fused decode chains**: attention runs as ONE 3-stage chain [qkv+bias/qk-norm/rope/kv-store] → [attn heads + per-head requant] → [wo+bias+residual], dense FFN as ONE 2-stage chain — 81 → 33 dispatches/token, +7.6% at the knee.

## dasLLAMA backend selection

- **Hybrid backend — per-slot batch pin** (`select_batch_backend` / `runtime.batch_backend`): decode (GEMV) and prefill (batch GEMM) can prefer different backends; on the EPYC tuned-portable wins decode while row-major acc8 wins batch (prefill 871 → 1044 t/s, decode unchanged). The pin overrides only batch-shaped slots from a layout-compatible donor and survives load-time re-selection.
- **AVX matrix tiers register unconditionally** — correctness held on real VNNI/AVX-512 silicon (probe + gpt-oss token-for-token, 4 backends), so the `DASLLAMA_AVX_MATRIX` master env gate is gone; registration is per-tier `cpu_supports` only, priorities unchanged (pin-only until the ISA ladder promotes a tier).
- **zvnni batch pin fleet-wide**: dense models 0.73–0.86× / gpt-oss 0.61× of llama.cpp prefill (+20% over unpinned), emit untouched. `emission_bench`/`prefill_perf` now apply the box_profile runtime before model load, making ttft a validated pp512 instrument (±1%).
- **M1 cross-check**: decode is backend-invariant on M1 (bandwidth-saturated) — confirms the frame: decode GEMV wants tuned-portable everywhere, batch GEMM wants the per-box batch tier; hybrid is an x64-only need today.

## Measurement hygiene

- `DAS_JOBQUE_THREADS` off-by-one documented (env sets WORKERS, lanes = N+1): fair vs `llama-bench -tN` is `DAS_JOBQUE_THREADS=N-1`; the 75%-of-lcpp headline held at true parity, and the final tables use it.
- Negative results logged rather than shipped: team-slot cache-line padding (4 layouts, all lose — the co-located slot is load-bearing), lane-cap surgery (dispatch tax was chunk compute, not protocol), whole-tensor repack (+3% only on this box).

## Docs + next-session plan

- New jobque builtins (`get_total_hw_cores`, `set_jobque_threads_cap`, `team_parallel_stages`/`_invoke`) wired into das2rst groups with handmade descriptions matching the `job_que.h` semantics.
- `modules/dasLLAMA/gemm_generator_plan.md` — the design + milestone runbook for the queued kernel-generator session: generate Q8/MXFP4 GEMM microkernels at the LLVM-IR layer (annotation-on-stub perm schema, tuned per-ISA × per-box × per-model × per-regime) instead of hand-writing the ~390-kernel family. Plan only; no code.

Tests: dasLLAMA suite 173/173 under `-jit` (on the EPYC box with all knobs active, and on arm64), new `tests/dasLLAMA/test_fused_decode.das` + `test_kernel_backend.das` + `tests/jobque/test_jobque_stages.das` (TSAN 25× clean). Full preflight green locally (interp + AOT + docs + sequence smoke); the JIT lane runs suite defaults in CI (the per-box `box_profile.json` is gitignored and never present there).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/misc/job_que.cpp [hunk 1]
```diff
@@ -83,10 +84,44 @@ namespace das {
...
+#endif
+
+    int JobQue::get_num_physical_cores() {
+#if defined(__APPLE__)
+        int ncores = 0; size_t len = sizeof(ncores);
+        if ( sysctlbyname("hw.physicalcpu", &ncores, &len, nullptr, 0) == 0 && ncores > 0 ) return ncores;
+#elif defined(_MSC_VER) && !defined(_GAMING_XBOX) && !defined(_DURANGO) && !defined(_M_ARM64)
+        if ( int ncores = GetPhysicalCoreCountInWindows() ) return ncores;
+#elif defined(__linux__)
+        // 2-way SMT is the universal x86 layout; smt/active is authoritative and cheap
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/job_que.cpp]: Good catch — fixed in f214619ae. `sysos.cpp` deliberately defines `GetPhysicalCoreCountInWindows` under `_WIN32` (its comment flags the old `_MSC_VER` gate as the bug that sent mingw to the fatal platform-fallback), so the `_MSC_VER` consumer guard here was exactly the mismatch you describe. Broadened both guard sites to `_WIN32 && !_GAMING_XBOX && !_DURANGO && !_M_ARM64` to match the definition. MSVC is unchanged (`_MSC_VER` implies `_WIN32`); MinGW/clang-cl now get the real physical count instead of the logical fallback.

File: src/misc/sysos.cpp [hunk 2]
```diff
@@ -54,6 +54,27 @@
...
+                    info = reinterpret_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX *>(buffer.data() + offset);
+                    if (info->Relationship == RelationProcessorCore) {
+                        coreCount++;    // one record per physical core, SMT siblings collapsed
+                    }
+                    offset += info->Size;
+                }
+                return coreCount;
+            }
+            return 0;
+        }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/sysos.cpp:76]: Leaving as-is — this is the established repo idiom (a verbatim twin of `GetLogicalProcessorCountInWindows` directly above it in the same file), and the flagged paths can't actually occur:

- **`info->Size == 0` / out-of-bounds loop:** the loop only runs when the second `GetLogicalProcessorInformationEx` returned `TRUE`, and a successful call has the OS fill valid, non-zero `Size` fields with in-bounds records — so `offset` always advances and stays within `returnLength`.
- **`buffer.resize(0)` from a failed probe:** the probe call with `RelationProcessorCore` only fails with `ERROR_INSUFFICIENT_BUFFER` (which sets `returnLength`); in the pathological `returnLength == 0` case, `resize(0)` → the second call sees a zero-length buffer, returns `FALSE`, and we hit `return 0`. No deref, no UB.

`return 0` is a safe fallback anyway — `get_num_physical_cores()` then returns the logical count and the thread cap stays optional. Adding guards here would only diverge from the sibling function for no reachable gain.

File: src/misc/job_que.cpp [hunk 3]
```diff
@@ -83,10 +84,44 @@ namespace das {
...
+        if ( FILE * f = fopen("/sys/devices/system/cpu/smt/active", "r") ) {
+            int active = 0;
+            int got = fscanf(f, "%d", &active);
+            fclose(f);
+            int logical = max(1, (int)thread::hardware_concurrency());
+            if ( got == 1 ) return active ? max(1, logical / 2) : logical;
+        }
+#endif
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/job_que.cpp:123]: Leaving as-is — the heuristic is correct on every platform this project actually builds for, and safe elsewhere:

- **x86 Linux** (the CI lane): 2-way SMT, `logical/2` is right.
- **ARM Linux** (the other CI lane, e.g. Graviton/Ampere): no SMT, so `/sys/.../smt/active` reads `0` → the code returns `logical`, correct.

The only case that returns a wrong count is genuine >2-way SMT on non-x86 (POWER SMT-4/8), which isn't a daslang target — and even there the result is a non-fatal perf-cap hint that falls back to a safe logical count, never a correctness bug. Deriving threads-per-core from full sysfs topology would be more portable, but it's added complexity for an untargeted platform; the code comment already scopes the reasoning to x86.

File: tests/dasLLAMA/test_kernel_backend.das [hunk 4]
```diff
@@ -57,3 +57,39 @@ def test_select_and_pin(t : T?) {
...
+        // an unknown donor warns and keeps the active backend's batch slots
+        select_batch_backend("no-such-backend")
+        t |> equal(active_batch_backend(), "portable", "unknown batch donor keeps active batch slots")
+        // a repack donor on a row-major active backend is layout-incompatible: warn + keep
+        for (n in kernel_backend_names()) {
+            if (n == "arm64-laneq" || n == "x64-avx2-repack") {
+                select_batch_backend(n)
+                t |> equal(active_batch_backend(), "portable", "repack donor rejected on row-major active")
+            }
+        }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests/dasLLAMA/test_kernel_backend.das:79]: Fixed in e33803787 — added a `tested_reject` tracker; when no repack donor (`arm64-laneq`/`x64-avx2-repack`) is registered, the test now emits `[W] ... layout-guard path not exercised` instead of silently passing. Verified: interp (portable-only) prints the skip note and passes; JIT arm64 finds `arm64-laneq`, exercises the reject path, no note. Same treatment applied to the sibling row-major loop below.

File: tests/dasLLAMA/test_kernel_backend.das [hunk 5]
```diff
@@ -57,3 +57,39 @@ def test_select_and_pin(t : T?) {
...
+        // a row-major donor takes the batch slots and survives re-activation
+        for (n in kernel_backend_names()) {
+            if (n == "arm64-sdot" || n == "x64-avx2-acc8") {
+                select_batch_backend(n)
+                t |> equal(active_batch_backend(), n, "row-major donor takes the batch slots")
+                select_kernel_backend("portable")   // re-activate: the standing pin must re-apply
+                t |> equal(active_batch_backend(), n, "batch pin survives re-selection")
+            }
+        }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests/dasLLAMA/test_kernel_backend.das:93]: Fixed in e33803787 — same pattern: a `tested_accept` tracker, and when no row-major donor (`arm64-sdot`/`x64-avx2-acc8`) is registered the test emits `[W] ... batch-pin override/survival not exercised`. Verified both tiers: interp prints both skip notes and passes; JIT arm64 finds `arm64-sdot`, exercises the override + survives-re-selection asserts, no note. CI now reflects what each lane actually exercised.

File: src/misc/job_que.cpp [hunk 6]
```diff
@@ -372,58 +419,123 @@ namespace das {
...
+                (*mTeamWorkS[s])(int(c), threadIndex);
+                mTeamRemainingS[s].fetch_sub(1, std::memory_order_release);
+                ran = true;
+            }
+            // stage barrier: stage s+1 reads what stage s wrote, so wait for ALL of stage s —
+            // the acquire pairs with every serving lane's release decrement.
+            while ( mTeamRemainingS[s].load(std::memory_order_acquire) != 0 ) {
+                if ( mShutdown.load(std::memory_order_relaxed) ) { s = nStages - 1; break; }
+                jobque_spin_relax();
+            }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/job_que.cpp:464]: Traced this — the deadlock can't occur, so leaving it as-is. The reasoning:

**A bailing worker never strands a chunk.** The `mShutdown` escape is only at the stage *barrier* (line 461), which a worker reaches after its claim loop (445-456). Inside that loop, `claim (fetch_add) → serve → decrement (fetch_sub)` has no shutdown check between the claim and the decrement — so any chunk a worker claimed is always decremented before it can bail. The escape skips only the *wait*, never a pending decrement.

**The caller is the guaranteed-progress lane.** `teamParallelForStages` serves every stage itself (517-524) with no shutdown check in its claim loop, so it claims + decrements every chunk no worker took, for every stage. Even if all workers bail at their barriers, the caller single-handedly drives `mTeamRemainingS[s]` to 0 and its barrier (528) proceeds. `mTeamOpS[s].fetch_add` hands out indices `0..K-1` exactly once, each paired with one decrement, so the counter always reaches 0 as long as one lane keeps pulling — and the caller always does.

**A publisher-side shutdown escape (the suggested alternative) would be a use-after-free.** `mTeamWorkS[s]` aliases the caller's stack (`stages[s].work`); if the caller returned early while workers were still dereferencing it, that's a dangling read. The asymmetry — workers may bail, the caller must finish — is deliberate and load-bearing.

Also note the single-stage `teamParallelFor` routes through this exact function, so this isn't new surface; and tearing the JobQue down from another thread while the owner is mid-dispatch is a lifecycle violation independent of this code. The stage-order storm test in `tests/jobque` (TSAN 25×) exercises the concurrent path.

File: daslib/jobque_boost.das [hunk 7]
```diff
@@ -585,6 +585,63 @@ def team_parallel_for(range_begin, range_end : int; num_jobs : int; var l : lamb
...
+    //! wrote. Each stage is ``int3(range_begin, range_end, num_chunks)``; the lambda receives
+    //! ``(stage, job_begin, job_end)`` with the same chunk split as ``team_parallel_for``. The whole
+    //! chain costs ONE publish + ONE join (the lambda is cloned once per worker), so a fused
+    //! matmul→matmul chain pays the dispatch tail once instead of per op. Falls back to running the
+    //! stages sequentially inline when team mode is off (or the pool/ranges are trivial).
+    var idx = 0
+    for (s in stages) {
+        invoke(l, idx, s.x, s.y)   // note, this is never called if job-que is there.
+        idx++
+    }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/jobque_boost.das:605]: Aligned the fallback in 993b2790d — it now guards the invoke with `s.y > s.x`, so a zero-range stage is a pure barrier just like `team_parallel_stages_invoke` (`actualChunks[s] == 0`).

One correction on the trigger, though: passing a *pre-declared* lambda doesn't reach this body — `TeamParallelStagesMacro.transform` runs `macro_verify(call.arguments[1] is ExprAscend, ...)`, so a non-inline lambda is a compile error ("expecting lambda declaration, ExprAscend"), not a silent fall-through. The `[tag_function_macro]` rewrites every *direct* call into `team_parallel_stages_invoke`, so this body is normally dead (the "note, this is never called if job-que is there" comment). It's reachable only through an indirect call via a function pointer — a corner case, but the placeholder should still mirror the real semantics, and now does. Verified `tests/jobque/test_jobque_stages.das` stays 5/5 under interp and JIT, and das2rst regenerates with no diff.

File: src/misc/job_que.cpp [hunk 8]
```diff
@@ -123,6 +158,7 @@ namespace das {
         , mThreadCount( 0 )
         , mJobsRunning(0) {
         mThreadCount = get_num_threads();
+        mTeamProf = getenv("DAS_TEAM_PROF") != nullptr;
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/job_que.cpp]: Fixed in ad754eb29 — now parses the value (`atoi(...) != 0`), so `DAS_TEAM_PROF=0` and empty are off, matching the documented `=1` contract and the sibling `DAS_JOBQUE_THREADS` in the same file. Verified: with `=0` the `[team-prof]` dump is suppressed; with `=1` it prints (`ops=9370 ...`); `tests/jobque` stays 5/5 both ways.

File: src/misc/job_que.cpp [hunk 9]
```diff
@@ -372,58 +419,123 @@ namespace das {
...
         return ran;
     }
 
     void JobQue::teamParallelFor ( int numChunks, const JobChunk & work ) {
+        TeamStage stage = { &work, numChunks };
+        teamParallelForStages(&stage, 1);
+    }
+
+    void JobQue::teamParallelForStages ( const TeamStage * stages, int numStages ) {
+        DAS_ASSERT(numStages >= 1 && numStages <= MAX_TEAM_STAGES);
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/misc/job_que.cpp:476]: Leaving the assert as-is — `numStages` can't be invalid here from any caller, so this is an internal invariant, not a boundary check:

- `teamParallelFor` (the single-op path) calls `teamParallelForStages(&stage, 1)` — a compile-time constant `1` and a non-null stack pointer.
- `team_parallel_stages_invoke` (the only other caller, the das builtin) validates *before* calling: `if (numStages <= 0) { …; return; }` and `if (numStages > MAX_TEAM_STAGES) context->throw_error_at(...)`, and passes a `TeamStage teamStages[MAX_TEAM_STAGES]` stack array (never null). So by the time it reaches `teamParallelForStages`, `1 <= numStages <= MAX_TEAM_STAGES` is guaranteed.

The user-facing validation (throwing a proper das error) already lives at that public boundary; the `DAS_ASSERT` is the codebase's idiom for a "can't happen" internal precondition. Adding a release guard here would put a branch in the core dispatch primitive (~160 calls per decoded token) for an unreachable case, and a silent early-return would *mask* a programming error the assert is meant to surface in debug — strictly worse than letting it fire.
